engine/src/savegame.cpp

```diff
@@ -174,6 +181,105 @@ void SaveFileCopy( const char *src, const char *dst )
     }
 }
 
+class Utf8Checker {
+public:
+    /*
+     * From https://helloacm.com/how-to-validate-utf-8-encoding-the-simple-utf-8-validation-algorithm/
+     */
+    bool validUtf8(const std::vector<BYTE> &data) {
+        for (long unsigned int i = 0; i < data.size(); ++ i) {
+            // 0xxxxxxx
+            int bit1 = (data[i] >> 7) & 1;
+            if (bit1 == 0) continue;
+            // 110xxxxx 10xxxxxx
+            int bit2 = (data[i] >> 6) & 1;
+            if (bit2 == 0) return false;
+            // 11
+            int bit3 = (data[i] >> 5) & 1;
+            if (bit3 == 0) {
+                // 110xxxxx 10xxxxxx
+                if ((++ i) < data.size()) {
+                    if (is10x(data[i])) {
+                        continue;
+                    }
+                    return false;
+                } else {
+                    return false;
+                }
+            }
+            int bit4 = (data[i] >> 4) & 1;
+            if (bit4 == 0) {
+                // 1110xxxx 10xxxxxx 10xxxxxx
+                if (i + 2 < data.size()) {
+                    if (is10x(data[i + 1]) && is10x(data[i + 2])) {
+                        i += 2;
+                        continue;
+                    }
+                    return false;
+                } else {
+                    return false;
+                }
+            }
+            int bit5 = (data[i] >> 3) & 1;
+            if (bit5 == 1) return false;
+            if (i + 3 < data.size()) {
+                if (is10x(data[i + 1]) && is10x(data[i + 2]) && is10x(data[i + 3])) {
+                    i += 3;
+                    continue;
+                }
+                return false;
+            } else {
+                return false;
+            }
+        }
+        return true;
+    }
+private:
+    inline bool is10x(int a) {
+        int bit1 = (a >> 7) & 1;
+        int bit2 = (a >> 6) & 1;
+        return (bit1 == 1) && (bit2 == 0);
+    }
+};
+
+std::vector<BYTE> readFile(std::string filename)
+/*
+ * From https://stackoverflow.com/a/21802936
+ */
+{
+    std::ifstream file(filename, std::ios::binary);
+    file.unsetf(std::ios::skipws);
+    std::streampos fileSize;
+
+    file.seekg(0, std::ios::end);
+    fileSize = file.tellg();
+    file.seekg(0, std::ios::beg);
+
+    std::vector<BYTE> vec;
+    vec.reserve(fileSize);
+
+    std::copy(std::istream_iterator<BYTE>(file),
+          std::istream_iterator<BYTE>(),
+          std::back_inserter(vec));
+
+    file.close();
+
+    return vec;
+}
+
+bool isUtf8SaveGame(std::string filename)
+{
+    std::string path = GetSaveDir() + filename;
+    std::vector<BYTE> savegame = readFile(path);
+    std::unique_ptr<Utf8Checker> check(new Utf8Checker);
```

Comment:
question: does this need to be a pointer to an instance? Or can it just be a local instance?
```
Utf8Checker check;
...
if (check.validUtf8(...
```
If we don't need to allocate, then let's not. If there's a reason we need to, then okay; but I think we may not need to - stack allocation should be sufficient.

NOTE: This wasn't quite clear before.

Reply:
Yes it can, @vegastrike/code-developers do you all agree on this?

Reply:
It could just be a local instance instead of a pointer, but personally, I don't think that it matters much one way or the other.

I say let's merge this PR as is.

Reply:
If the checker is only ever used inside the scope of the isUtf8SaveGame function, then it makes sense to just use it directly since it doesn't need to be passed around to other consumers?

Reply:
I thought we could save some time by leaving the code the way it is here, but I see that that is not the way this is going. So sure, let's change it to use the UTF8Checker directly, as a local instance.